src/stream/src/common/table/state_table.rs

```diff
@@ -848,18 +1011,30 @@ where
                 }
             }
         }
+        self.prev_cleaned_watermark = watermark;
+
+        // Clear the watermark cache and force a resync.
+        if USE_WATERMARK_CACHE && !delete_ranges.is_empty() {
+            self.watermark_cache.clear();
+        }
```

Comment:
If the largest end of all delete ranges never larger than the largest value in the cache, it seems we can avoid clearing here? We'll need a `cache.drain_until(smallest_kept)` or a general range delete interface then.

Reply:
I had considered it previously. Seems to make the code more complicated, so I preferred to just clear the cache, as I don't think it would have too much performance impact.

We have to:
1. add the `drain_until` interface.
2. Decide whether to still refill the cache on barrier. If we do, we need to add logic to scan from `[highest_value_in_cache, +inf)`.

I will leave a note that this can be optimized in the future if necessary.